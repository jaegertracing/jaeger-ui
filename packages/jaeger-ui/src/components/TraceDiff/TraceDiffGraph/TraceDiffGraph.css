/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.TraceDiffGraph--graphWrapper {
  bottom: 0;
  cursor: move;
  left: 0;
  overflow: auto;
  position: absolute;
  right: 0;
  top: 0;
}

.TraceDiffGraph--errorsWrapper {
  background: #eee;
  bottom: 0;
  left: 0;
  overflow: auto;
  padding: 5rem 3.5rem;
  position: absolute;
  right: 0;
  top: 0;
}

.TraceDiffGraph--errorMessage {
  font-size: 1.8rem;
}

.TraceDiffGraph--dag {
  transition: background 0.5s ease;
  stroke-width: 0.9;
  z-index: -1;
}

.TraceDiffGraph--dag.is-uiFind-mode {
  background: #ddd;
}

/* Find within diff */
.TraceDiffGraph--uiFind {
  bottom: 20px;
  position: absolute;
  right: 20px;
  width: 300px;
  z-index: 1;
}

.TraceDiffGraph--uiFind > .ant-input {
  border: 1px solid #aaa;
  border-radius: 0;
  box-shadow: 0 0 2px 2px #ccc;
}
