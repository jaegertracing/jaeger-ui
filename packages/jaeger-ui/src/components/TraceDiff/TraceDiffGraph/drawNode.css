/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.DiffNode {
  background: #bbb;
  border: 1px solid #777;
  box-shadow: 0 0px 3px rgba(0, 0, 0, 0.2);
  cursor: pointer;
  white-space: nowrap;
}

.TraceDiffGraph--dag.is-small .DiffNode > tbody {
  opacity: 0;
}

.DiffNode.is-changed {
  color: var(--tx-color-title);
}

.DiffNode.is-more {
  background: #78d539;
  border-color: #2a8f04;
}

.DiffNode.is-added {
  background: #2a8f04;
  border: none;
  color: #fff;
}

.DiffNode.is-less {
  border-color: #cc1616;
  background: #ffa39e;
}

.DiffNode.is-removed {
  background: #cc1616;
  border: none;
  color: #fff;
}

.DiffNode--metricCell {
  padding: 0.3rem 0.5rem;
  background: rgba(255, 255, 255, 0.3);
}

.DiffNode--metricCell.is-changed {
  font-weight: 500;
}

.DiffNode--metricCell.is-added,
.DiffNode--metricCell.is-removed {
  background: rgba(0, 0, 0, 0.2);
}

.DiffNode--metricSymbol {
  margin: 0 0.1em;
}

.DiffNode--labelCell {
  padding: 0.3rem 0.5rem 0.3rem 0.75rem;
}

/* Tweak the popover aesthetics - unfortunate but necessary */

.DiffNode--popover .ant-popover-inner-content {
  padding: 0;
  position: relative;
}

.DiffNode--popover.is-same .ant-popover-arrow {
  background: #777;
}

.DiffNode--popover.is-more .ant-popover-arrow,
.DiffNode--popover.is-added .ant-popover-arrow {
  background: #2a8f04;
}

.DiffNode--popover.is-less .ant-popover-arrow,
.DiffNode--popover.is-removed .ant-popover-arrow {
  background: #cc1616;
}
