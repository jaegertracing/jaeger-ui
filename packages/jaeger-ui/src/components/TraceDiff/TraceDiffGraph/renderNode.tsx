// Copyright (c) 2018-2020 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Popover } from 'antd';
import cx from 'classnames';
import { TLayoutVertex } from '@jaegertracing/plexus/lib/types';

import EmphasizedNode from '../../common/EmphasizedNode';
import CopyIcon from '../../common/CopyIcon';
import { TDiffCounts } from '../../../model/trace-dag/types';
import TDagPlexusVertex from '../../../model/trace-dag/types/TDagPlexusVertex';

import './renderNode.css';

type Props = {
  a: number;
  b: number;
  operation: string;
  service: string;
};

const abs = Math.abs;
const max = Math.max;

export class DiffNode extends React.PureComponent<Props> {
  render() {
    const { a, b, operation, service } = this.props;
    const isSame = a === b;
    const className = cx({
      'is-same': isSame,
      'is-changed': !isSame,
      'is-more': b > a && a > 0,
      'is-added': a === 0,
      'is-less': a > b && b > 0,
      'is-removed': b === 0,
    });
    const chgSign = a < b ? '+' : '-';
    const table = (
      <table className={`DiffNode ${className}`}>
        <tbody className="DiffNode--body">
          <tr>
            <td className={`DiffNode--metricCell ${className}`} rowSpan={isSame ? 2 : 1}>
              {isSame ? null : <span className="DiffNode--metricSymbol">{chgSign}</span>}
              {isSame ? a : abs(b - a)}
            </td>
            <td className={`DiffNode--labelCell ${className}`}>
              <strong>{service}</strong>
              <CopyIcon
                className="DiffNode--copyIcon"
                copyText={`${service} ${operation}`}
                tooltipTitle="Copy label"
              />
            </td>
          </tr>
          <tr>
            {isSame ? null : (
              <td className={`DiffNode--metricCell ${className}`}>
                <span className="DiffNode--metricSymbol">{chgSign}</span>
                {a === 0 || b === 0 ? 100 : abs(((a - b) / max(a, b)) * 100).toFixed(0)}
                <span className="DiffNode--metricSymbol">%</span>
              </td>
            )}
            <td className={`DiffNode--labelCell ${className}`}>{operation}</td>
          </tr>
        </tbody>
      </table>
    );

    return (
      <Popover overlayClassName={`DiffNode--popover ${className}`} mouseEnterDelay={0.25} content={table}>
        {table}
      </Popover>
    );
  }
}

export default function renderNode(vertex: TDagPlexusVertex<TDiffCounts>) {
  const { a, b, operation, service } = vertex.data;
  const lenA = a ? a.length : 0;
  const lenB = b ? b.length : 0;
  return <DiffNode a={lenA} b={lenB} operation={operation} service={service} />;
}

export function getNodeEmphasisRenderer(keys: Set<string>) {
  return function drawEmphasizedNode(lv: TLayoutVertex<any>) {
    if (!keys.has(lv.vertex.key)) {
      return null;
    }
    return <EmphasizedNode height={lv.height} width={lv.width} />;
  };
}
