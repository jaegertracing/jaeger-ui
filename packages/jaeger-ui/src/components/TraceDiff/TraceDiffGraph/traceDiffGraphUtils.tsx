// Copyright (c) 2019-2020 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { TVertexKey } from '@jaegertracing/plexus/lib/types';
import _get from 'lodash/get';
import _map from 'lodash/map';
import memoizeOne from 'memoize-one';

import convPlexus from '../../../model/trace-dag/convPlexus';
import TraceDag from '../../../model/trace-dag/TraceDag';
import { TDenseSpanMembers, TDiffCounts } from '../../../model/trace-dag/types';
import TDagPlexusVertex from '../../../model/trace-dag/types/TDagPlexusVertex';
import { Trace } from '../../../types/trace';
import filterSpans from '../../../utils/filter-spans';

function getUiFindVertexKeysFn(
  uiFind: string,
  vertices: TDagPlexusVertex<TDenseSpanMembers>[]
): Set<TVertexKey> {
  if (!uiFind) return new Set<TVertexKey>();
  const newVertexKeys: Set<TVertexKey> = new Set();
  vertices.forEach(({ key, data: { members } }) => {
    if (_get(filterSpans(uiFind, _map(members, 'span')), 'size')) {
      newVertexKeys.add(key);
    }
  });
  return newVertexKeys;
}

export const getUiFindVertexKeys = memoizeOne(getUiFindVertexKeysFn);

function getEdgesAndVerticesFn(aData: Trace, bData: Trace) {
  const aTraceDag = TraceDag.newFromTrace(aData);
  const bTraceDag = TraceDag.newFromTrace(bData);
  const diffDag = TraceDag.diff(aTraceDag, bTraceDag);
  return convPlexus<TDiffCounts>(diffDag.nodesMap);
}

export const getEdgesAndVertices = memoizeOne(getEdgesAndVerticesFn);
