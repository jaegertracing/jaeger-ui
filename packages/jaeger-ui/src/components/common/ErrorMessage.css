/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.ErrorMessage {
  position: relative;
  word-break: break-word;
  word-wrap: break-word;
}

.ErrorMessage--msg {
  color: #c00;
  margin: 0;
  padding-bottom: 1rem;
}

.ErrorMessage--details {
  background: #f5f5f5;
  border: 1px solid #e6e6e6;
  overflow: auto;
}

.ErrorMessage--detailsTable {
  min-width: 100%;
}

.ErrorMessage--detailItem:nth-child(2n) > .ErrorMessage--attr,
.ErrorMessage--detailItem:nth-child(2n) > .ErrorMessage--value {
  background-color: #f0f0f0;
  border-bottom: 1px solid #e8e8e8;
  border-top: 1px solid #e8e8e8;
}

/* dont have extra borders on the bottom row */
.ErrorMessage--detailItem:last-child > .ErrorMessage--attr,
.ErrorMessage--detailItem:last-child > .ErrorMessage--value {
  border-bottom: none;
}

.ErrorMessage--attr {
  color: rgba(0, 0, 0, 0.45);
  padding: 0.5em;
  vertical-align: top;
  white-space: nowrap;
}

.ErrorMessage--value {
  color: rgba(0, 0, 0, 0.8);
  font-family: monospace;
  padding: 0.5em;
  white-space: pre;
}
