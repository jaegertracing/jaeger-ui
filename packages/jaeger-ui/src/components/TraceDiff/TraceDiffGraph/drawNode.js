// @flow

// Copyright (c) 2018 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Popover } from 'antd';
import cx from 'classnames';

import CopyIcon from '../../common/CopyIcon';

import type { PVertex } from '../../../model/trace-dag/types';

import './drawNode.css';

type Props = {
  a: number,
  b: number,
  operation: string,
  service: string,
};

const abs = Math.abs;
const max = Math.max;

class DiffNode extends React.PureComponent<Props> {
  props: Props;

  render() {
    const { a, b, operation, service } = this.props;
    const isSame = a === b;
    const className = cx({
      'is-same': isSame,
      'is-changed': !isSame,
      'is-more': b > a && a > 0,
      'is-added': a === 0,
      'is-less': a > b && b > 0,
      'is-removed': b === 0,
    });
    const chgSign = a < b ? '+' : '-';
    const table = (
      <table className={`DiffNode ${className}`}>
        <tbody>
          <tr>
            <td className={`DiffNode--metricCell ${className}`} rowSpan={isSame ? 2 : 1}>
              {isSame ? null : <span className="DiffNode--metricSymbol">{chgSign}</span>}
              {isSame ? a : abs(b - a)}
            </td>
            <td className={`DiffNode--labelCell ${className}`}>
              <strong>{service}</strong>
              <CopyIcon
                className="DiffNode--copyIcon"
                copyText={`${service} ${operation}`}
                tooltipTitle="Copy label"
              />
            </td>
          </tr>
          <tr>
            {isSame ? null : (
              <td className={`DiffNode--metricCell ${className}`}>
                <span className="DiffNode--metricSymbol">{chgSign}</span>
                {a === 0 || b === 0 ? 100 : abs((a - b) / max(a, b) * 100).toFixed(0)}
                <span className="DiffNode--metricSymbol">%</span>
              </td>
            )}
            <td className={`DiffNode--labelCell ${className}`}>{operation}</td>
          </tr>
        </tbody>
      </table>
    );

    return (
      <Popover overlayClassName={`DiffNode--popover ${className}`} mouseEnterDelay={0.25} content={table}>
        {table}
      </Popover>
    );
  }
}

export default function drawNode<T>(vertex: PVertex<T>) {
  const { data, operation, service } = vertex.data;
  return <DiffNode {...data} operation={operation} service={service} />;
}
