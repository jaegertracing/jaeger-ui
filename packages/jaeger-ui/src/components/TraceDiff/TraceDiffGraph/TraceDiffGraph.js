// @flow

// Copyright (c) 2018 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { DirectedGraph, LayoutManager } from '@jaegertracing/plexus';

import drawNode from './drawNode';
import ErrorMessage from '../../common/ErrorMessage';
import LoadingIndicator from '../../common/LoadingIndicator';
import { fetchedState } from '../../../constants';
import convPlexus from '../../../model/trace-dag/convPlexus';
import TraceDag from '../../../model/trace-dag/TraceDag';

import type { FetchedTrace } from '../../../types';

import './TraceDiffGraph.css';

type Props = {
  a: ?FetchedTrace,
  b: ?FetchedTrace,
};

const { classNameIsSmall } = DirectedGraph.propsFactories;

function setOnEdgesContainer(state: Object) {
  const { zoomTransform } = state;
  if (!zoomTransform) {
    return null;
  }
  const { k } = zoomTransform;
  const opacity = 0.1 + k * 0.9;
  return { style: { opacity } };
}

export default class TraceDiffGraph extends React.PureComponent<Props> {
  props: Props;

  layoutManager: LayoutManager;

  constructor(props: Props) {
    super(props);
    this.layoutManager = new LayoutManager({
      useDotEdges: true,
      splines: 'polyline',
    });
  }

  componentWillUnmount() {
    this.layoutManager.stopAndRelease();
  }

  render() {
    const { a, b } = this.props;
    if (!a || !b) {
      return <h1 className="u-mt-vast u-tx-muted ub-tx-center">At least two Traces are needed</h1>;
    }
    if (a.error || b.error) {
      return (
        <div className="TraceDiffGraph--errorsWrapper">
          {a.error && (
            <ErrorMessage
              className="ub-my4"
              error={a.error}
              messageClassName="TraceDiffGraph--errorMessage"
            />
          )}
          {b.error && (
            <ErrorMessage
              className="ub-my4"
              error={b.error}
              messageClassName="TraceDiffGraph--errorMessage"
            />
          )}
        </div>
      );
    }
    if (a.state === fetchedState.LOADING || b.state === fetchedState.LOADING) {
      return <LoadingIndicator className="u-mt-vast" centered />;
    }
    const aData = a.data;
    const bData = b.data;
    if (!aData || !bData) {
      return <div className="TraceDiffGraph--graphWrapper" />;
    }
    const aTraceDag = TraceDag.newFromTrace(aData);
    const bTraceDag = TraceDag.newFromTrace(bData);
    const diffDag = TraceDag.diff(aTraceDag, bTraceDag);
    const { edges, vertices } = convPlexus(diffDag.nodesMap);

    return (
      <div className="TraceDiffGraph--graphWrapper">
        <DirectedGraph
          minimap
          zoom
          arrowScaleDampener={0}
          className="TraceDiffGraph--dag"
          minimapClassName="TraceDiffGraph--miniMap"
          layoutManager={this.layoutManager}
          getNodeLabel={drawNode}
          setOnRoot={classNameIsSmall}
          setOnEdgesContainer={setOnEdgesContainer}
          edges={edges}
          vertices={vertices}
        />
      </div>
    );
  }
}
