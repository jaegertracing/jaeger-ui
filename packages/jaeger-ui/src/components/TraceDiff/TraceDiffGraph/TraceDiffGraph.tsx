// Copyright (c) 2018 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { cacheAs, Digraph, LayoutManager } from '@jaegertracing/plexus';
import cx from 'classnames';
import { connect } from 'react-redux';

import renderNode, { getNodeEmphasisRenderer } from './renderNode';
import { getUiFindVertexKeys, getEdgesAndVertices } from './traceDiffGraphUtils';
import ErrorMessage from '../../common/ErrorMessage';
import LoadingIndicator from '../../common/LoadingIndicator';
import UiFindInput, { extractUiFindFromState, TExtractUiFindFromStateReturn } from '../../common/UiFindInput';
import { fetchedState } from '../../../constants';
import { FetchedTrace, TNil } from '../../../types';

import './TraceDiffGraph.css';

type Props = {
  a: FetchedTrace | TNil;
  b: FetchedTrace | TNil;
} & TExtractUiFindFromStateReturn;

const { classNameIsSmall, scaleOpacity, scaleStrokeOpacity } = Digraph.propsFactories;

export class UnconnectedTraceDiffGraph extends React.PureComponent<Props> {
  layoutManager = new LayoutManager({ useDotEdges: true, splines: 'polyline' });

  cacheAs = cacheAs.makeScope();

  componentWillUnmount() {
    this.layoutManager.stopAndRelease();
  }

  render() {
    const { a, b, uiFind = '' } = this.props;
    if (!a || !b) {
      return <h1 className="u-mt-vast u-tx-muted ub-tx-center">At least two Traces are needed</h1>;
    }
    if (a.error || b.error) {
      return (
        <div className="TraceDiffGraph--errorsWrapper">
          {a.error && (
            <ErrorMessage
              className="ub-my4"
              error={a.error}
              messageClassName="TraceDiffGraph--errorMessage"
            />
          )}
          {b.error && (
            <ErrorMessage
              className="ub-my4"
              error={b.error}
              messageClassName="TraceDiffGraph--errorMessage"
            />
          )}
        </div>
      );
    }
    if (a.state === fetchedState.LOADING || b.state === fetchedState.LOADING) {
      return <LoadingIndicator className="u-mt-vast" centered />;
    }
    const aData = a.data;
    const bData = b.data;
    if (!aData || !bData) {
      return <div className="TraceDiffGraph--graphWrapper" />;
    }
    const { edges, vertices } = getEdgesAndVertices(aData, bData);
    const keys = getUiFindVertexKeys(uiFind, vertices);
    const dagClassName = cx('TraceDiffGraph--dag', { 'is-uiFind-mode': uiFind });
    const inputProps: Record<string, string | undefined> = {
      className: 'TraceDiffGraph--uiFind',
      suffix: uiFind.length ? String(keys.size) : undefined,
    };

    return (
      <div className="TraceDiffGraph--graphWrapper">
        <Digraph
          // `key` is necessary to see updates to the graph when a or b changes
          // TODO(joe): debug this issue in Digraph
          key={`${a.id} vs ${b.id}`}
          minimap
          zoom
          className={dagClassName}
          minimapClassName="u-miniMap"
          layoutManager={this.layoutManager}
          measurableNodesKey="nodes"
          layers={[
            {
              key: 'emphasis-nodes',
              layerType: 'svg',
              renderNode: getNodeEmphasisRenderer(keys),
            },
            {
              key: 'edges',
              layerType: 'svg',
              edges: true,
              defs: [{ localId: 'arrow' }],
              markerEndId: 'arrow',
              setOnContainer: this.cacheAs('edges/container', [
                scaleOpacity,
                scaleStrokeOpacity,
                { stroke: '#444' },
              ]),
            },
            {
              renderNode,
              key: 'nodes',
              measurable: true,
              layerType: 'html',
            },
          ]}
          setOnGraph={classNameIsSmall}
          edges={edges}
          vertices={vertices}
        />
        <UiFindInput inputProps={inputProps} />
      </div>
    );
  }
}

export default connect(extractUiFindFromState)(UnconnectedTraceDiffGraph);
