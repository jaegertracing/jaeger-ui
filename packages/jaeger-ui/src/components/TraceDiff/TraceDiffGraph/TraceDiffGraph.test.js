// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { shallow } from 'enzyme';

import { UnconnectedTraceDiffGraph as TraceDiffGraph } from './TraceDiffGraph';
import ErrorMessage from '../../common/ErrorMessage';
import LoadingIndicator from '../../common/LoadingIndicator';
import UiFindInput from '../../common/UiFindInput';
import { fetchedState } from '../../../constants';

describe('TraceDiffGraph', () => {
  const props = {
    a: {
      data: {
        spans: [],
        traceID: 'trace-id-a',
      },
      error: null,
      id: 'trace-id-a',
      state: fetchedState.DONE,
    },
    b: {
      data: {
        spans: [],
        traceID: 'trace-id-b',
      },
      error: null,
      id: 'trace-id-b',
      state: fetchedState.DONE,
    },
  };
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<TraceDiffGraph {...props} />);
  });

  it('renders warning when a or b are not provided', () => {
    expect(wrapper.find('h1').length).toBe(0);

    wrapper.setProps({ a: undefined });
    expect(wrapper.find('h1').length).toBe(1);
    expect(wrapper.find('h1').text()).toBe('At least two Traces are needed');

    wrapper.setProps({ b: undefined });
    expect(wrapper.find('h1').length).toBe(1);
    expect(wrapper.find('h1').text()).toBe('At least two Traces are needed');

    wrapper.setProps({ a: props.a });
    expect(wrapper.find('h1').length).toBe(1);
    expect(wrapper.find('h1').text()).toBe('At least two Traces are needed');
  });

  it('renders warning when a or b have errored', () => {
    expect(wrapper.find(ErrorMessage).length).toBe(0);

    const errorA = 'some error text for trace a';
    wrapper.setProps({
      a: {
        ...props.a,
        error: errorA,
      },
    });

    expect(wrapper.find(ErrorMessage).length).toBe(1);
    expect(wrapper.find(ErrorMessage).props()).toEqual(
      expect.objectContaining({
        error: errorA,
      })
    );
    const errorB = 'some error text for trace a';
    wrapper.setProps({
      b: {
        ...props.b,
        error: errorB,
      },
    });

    expect(wrapper.find(ErrorMessage).length).toBe(2);
    expect(
      wrapper
        .find(ErrorMessage)
        .at(1)
        .props()
    ).toEqual(
      expect.objectContaining({
        error: errorB,
      })
    );
    wrapper.setProps({
      a: props.a,
    });
    expect(wrapper.find(ErrorMessage).length).toBe(1);
    expect(wrapper.find(ErrorMessage).props()).toEqual(
      expect.objectContaining({
        error: errorB,
      })
    );
  });

  it('renders a loading indicator when a or b are loading', () => {
    expect(wrapper.find(LoadingIndicator).length).toBe(0);

    wrapper.setProps({
      a: {
        state: fetchedState.LOADING,
      },
    });
    expect(wrapper.find(LoadingIndicator).length).toBe(1);

    wrapper.setProps({
      b: {
        state: fetchedState.LOADING,
      },
    });
    expect(wrapper.find(LoadingIndicator).length).toBe(1);

    wrapper.setProps({ a: props.a });
    expect(wrapper.find(LoadingIndicator).length).toBe(1);
  });

  it('renders an empty div when a or b lack data', () => {
    expect(wrapper.children().length).not.toBe(0);

    const { data: unusedAData, ...aWithoutData } = props.a;
    wrapper.setProps({ a: aWithoutData });
    expect(wrapper.children().length).toBe(0);

    const { data: unusedBData, ...bWithoutData } = props.b;
    wrapper.setProps({ b: bWithoutData });
    expect(wrapper.children().length).toBe(0);

    wrapper.setProps({ a: props.a });
    expect(wrapper.children().length).toBe(0);
  });

  it('renders a DiGraph when it has data', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders current uiFind count when given uiFind', () => {
    expect(wrapper.find(UiFindInput).prop('inputProps')).toEqual(
      expect.objectContaining({
        suffix: undefined,
      })
    );

    wrapper.setProps({ uiFind: 'test uiFind' });

    expect(wrapper.find(UiFindInput).prop('inputProps')).toEqual(
      expect.objectContaining({
        suffix: '0',
      })
    );
  });

  it('cleans up layoutManager before unmounting', () => {
    const layoutManager = jest.spyOn(wrapper.instance().layoutManager, 'stopAndRelease');
    wrapper.unmount();
    expect(layoutManager).toHaveBeenCalledTimes(1);
  });
});
