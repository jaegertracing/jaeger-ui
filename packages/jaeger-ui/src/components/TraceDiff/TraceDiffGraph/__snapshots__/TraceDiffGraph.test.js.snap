// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TraceDiffGraph renders a DiGraph when it has data 1`] = `
<div
  className="TraceDiffGraph--graphWrapper"
>
  <Digraph
    className="TraceDiffGraph--dag"
    classNamePrefix="plexus"
    edges={Array []}
    key="trace-id-a vs trace-id-b"
    layers={
      Array [
        Object {
          "key": "emphasis-nodes",
          "layerType": "svg",
          "renderNode": [Function],
        },
        Object {
          "defs": Array [
            Object {
              "localId": "arrow",
            },
          ],
          "edges": true,
          "key": "edges",
          "layerType": "svg",
          "markerEndId": "arrow",
          "setOnContainer": Array [
            [Function],
            [Function],
            Object {
              "stroke": "#444",
            },
          ],
        },
        Object {
          "key": "nodes",
          "layerType": "html",
          "measurable": true,
          "renderNode": [Function],
        },
      ]
    }
    layoutManager={
      LayoutManager {
        "_handleUpdate": [Function],
        "coordinator": Coordinator {
          "_handleVizWorkerError": [Function],
          "_handleVizWorkerMessage": [Function],
          "_handleVizWorkerMessageError": [Function],
          "busyWorkers": Array [],
          "callback": [Function],
          "currentLayout": null,
          "idleWorkers": Array [],
          "nextWorkerId": 0,
        },
        "layoutId": 0,
        "options": Object {
          "splines": "polyline",
          "useDotEdges": true,
        },
        "pendingResult": null,
      }
    }
    measurableNodesKey="nodes"
    minimap={true}
    minimapClassName="u-miniMap"
    setOnGraph={[Function]}
    vertices={Array []}
    zoom={true}
  />
  <Connect(WithRouteProps)
    inputProps={
      Object {
        "className": "TraceDiffGraph--uiFind",
        "suffix": undefined,
      }
    }
  />
</div>
`;
